src/runtime/runtime_unix.go

```diff
@@ -222,46 +224,30 @@ func nanosecondsToTicks(ns int64) timeUnit {
 }
 
 func sleepTicks(d timeUnit) {
-	// When there are no signal handlers present, we can simply go to sleep.
-	if !hasSignals {
-		// timeUnit is in nanoseconds, so need to convert to microseconds here.
-		usleep(uint(d) / 1000)
-		return
+	var until timeUnit
+	if !hasScheduler {
+		until = ticks() + d
 	}
 
-	if GOOS == "darwin" {
-		// Check for incoming signals.
-		if checkSignals() {
-			// Received a signal, so there's probably at least one goroutine
-			// that's runnable again.
-			return
+	for {
+		// Sleep for the given amount of time.
+		// If a signal arrived before going to sleep, or during the sleep, the
+		// sleep will exit early.
+		signalFutex.WaitUntil(0, uint64(ticksToNanoseconds(d)))
+
+		// Check whether there was a signal before or during the call to
+		// WaitUntil.
+		if signalFutex.Swap(0) != 0 {
+			if checkSignals() && hasScheduler {
+				return
+			}
 		}
 
-		// WARNING: there is a race condition here. If a signal arrives between
-		// checkSignals() and usleep(), the usleep() call will not exit early so
-		// the signal is delayed until usleep finishes or another signal
-		// arrives.
-		// There doesn't appear to be a simple way to fix this on MacOS.
-
-		// timeUnit is in nanoseconds, so need to convert to microseconds here.
-		result := usleep(uint(d) / 1000)
-		if result != 0 {
-			checkSignals()
-		}
-	} else {
-		// Linux (and various other POSIX systems) implement sigtimedwait so we
-		// can do this in a non-racy way.
-		tinygo_wfi_mask(activeSignals)
-		if checkSignals() {
-			tinygo_wfi_unmask()
+		// Set duration (in next loop iteration) to the remaining time.
+		d = until - ticks()
+		if d <= 0 {
```

Comment:
If `hasScheduler`, d will always be less than zero, no?

Reply:
Not necessarily! There are cases where `Wait` and `WaitUntil` might return due to spurious wakeups and not due to a signal. I'm not entirely sure whether this is possible, but the futex documentation on Linux is a bit vague on it (see src/internal/futex/futex.go) and since there's no real documentation for the MacOS syscalls I think it's better to defensively assume they can return spuriously. (I know that on Windows, the equivalent syscalls do return spuriously sometimes).